### PR TITLE
SCP-1857: Add some types for the new contract instance model in the PAB

### PR DESCRIPTION
**Pull request by j-mueller**
* Add some types for the new contract instance model in the PAB

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
- PR
    - [x] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested
- If you updated any cabal files or added Haskell packages:
    - [ ] `nix-shell shell.nix --run updateMaterialized` to update the materialized Nix files
    - [ ] Update `hie-*.yaml` files if needed
- If you changed any Haskell files:
    - x ] `nix-shell shell.nix --run fix-stylish-haskell` to fix any formatting issues
- If you changed any Purescript files:
    - [ ] `nix-shell shell.nix --run fix-purty` to fix any formatting issues

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

**Comment by j-mueller** [comment]
This design gets rid of constant polling, follows the principles from Duncan's talk, and can (eventually) be implemented in the emulator with relatively few changes as well.

Besides the STM / waiting architecture the new idea is that it deals with rollbacks at the level of individual transactions. That way we have some freedom in the contracts to decide how to react to those rollbacks that affect us.

**Review by raduom — APPROVED** [review]
LGTM with the caveat that I don't really see how this is going to get used.

**Comment by j-mueller** [comment]
> how this is going to get used

It will replace calling `processContractOutboxes` every second

File: plutus-pab/src/Plutus/PAB/Core/ContractInstance/STM.hs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+# QoS
+
+One of the main goals of the queueless STM design is to avoid a degradation
+of the quality of service (QoS) when the system is operating at capacity. This
+comes at a price: When the system is under pressure, some updates may be
+dropped. In practice this a result of the behaviour of STM's 'retry' primitive,
+which only guarantees to retry at some point (not immediately) after a variable
+has changed. So if the variable changes again before the retry happens, the
+intermediate state is not visible.
```
Comment (j-mueller) [inline plutus-pab/src/Plutus/PAB/Core/ContractInstance/STM.hs:50]: It should also be noted that we can assume the PAB is running in a friendly environment, ie. it is much less likely to be the target of DoS attacks than the node is. In fact the node is the most likely source of increased traffic for us. So as long as we can deal with the incoming transactions from the node fast enough, we can be a little more relaxed about QoS in other places.

File: plutus-pab/src/Plutus/PAB/Core/ContractInstance/STM.hs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+to be notified of new states. Therefore we choose the following types for the
+events that we want to know about.
+
+* Time: TVar with current time
+* Modifications to an address: TVar with current UTXO set of that address
+* Transactions & rollbacks: TVar with current status of transactions
+* Endpoints: For each endpoint a TMVar that changes from empty to full if & when
+  the endpoint gets called.
+
+All other requests of the contract are handled using the wallet effects (chain index,
```
Comment (gilligan) [inline plutus-pab/src/Plutus/PAB/Core/ContractInstance/STM.hs:64]: nice job on the documentation - well written.

**Review by gilligan — APPROVED** [review]

**Pull request merged**